server/schedule/region_splitter.go

```diff
@@ -119,3 +136,54 @@ func (r *RegionSplitter) groupKeysByRegion(keys [][]byte) (map[uint64][][]byte,
 	}
 	return groupKeys, unProcessedKeys
 }
+
+func (r *RegionSplitter) checkRegionValid(region *core.RegionInfo) bool {
+	if r.cluster.IsRegionHot(region) {
+		return false
+	}
+	if !opt.IsRegionReplicated(r.cluster, region) {
+		r.cluster.AddSuspectRegions(region.GetID())
+		return false
+	}
+	if region.GetLeader() == nil {
+		return false
+	}
+	return true
+}
+
+type splitRegionsHandler struct {
+	cluster opt.Cluster
+	oc      *OperatorController
+}
+
+func (h *splitRegionsHandler) SplitRegionByKeys(region *core.RegionInfo, splitKeys [][]byte) error {
+	op := operator.CreateSplitRegionOperator("region-splitter", region, 0, pdpb.CheckPolicy_USEKEY, splitKeys)
+	if ok := h.oc.AddOperator(op); !ok {
+		log.Info("add region split operator failed", zap.Uint64("regionID", region.GetID()))
+		return errors.New("add region split operator failed")
+	}
+	return nil
+}
+
+func (h *splitRegionsHandler) WatchRegionsByKeyRange(startKey, endKey []byte, expectRegionsCount int, timeout, watchInterval time.Duration) map[uint64]struct{} {
+	after := time.After(timeout)
+	ticker := time.NewTicker(watchInterval)
+	defer ticker.Stop()
+	var regionsID map[uint64]struct{}
+	for {
+		select {
+		case <-ticker.C:
+			regions := h.cluster.ScanRegions(startKey, endKey, expectRegionsCount)
+			regionsID = make(map[uint64]struct{}, expectRegionsCount)
+			for _, region := range regions {
+				regionsID[region.GetID()] = struct{}{}
+			}
+			if len(regionsID) < expectRegionsCount {
```

Comment:
Should we consider that a region may split due to other reason (e.g. oversize)? In this case, even if region count is enough, the split keys may different from expected.

Reply:
That's a good point. I think we can watch and asserting the regions by comparing split keys and regions' start key.